c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,117 @@ fn run(tcx: TyCtxt) {
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+
+                    // Finds casts associated with the special functions malloc, calloc,
+                    // realloc, and free. For the first three, there is expected to be
+                    // a cast from void* to an arbitrary type in the subsequent block.
+                    // For realloc and free, there is expected to be a cast to void*
+                    // from an arbitrary type in the same block.
+                    let mut handle_c_void_cast = |void_ptr: &Place<'tcx>,
+                                                  find_cast: &dyn Fn() -> Option<(
+                        Place<'tcx>,
+                        Place<'tcx>,
+                    )>| {
+                        let deref_ty = void_ptr
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+                        });
+
+                        if let Some((casted, castee)) = find_cast() {
+                            if castee == *void_ptr {
+                                // This is a special case for types being casted from *c_void to a pointer
+                                // to some other type, e.g. `let foo = malloc(..) as *mut Foo;`
+                                // carry over the pointer id of *c_void, but match the pointer ids
+                                // of the casted-to-type for the rest
+                                acx.c_void_casts.0.insert(castee, casted);
+                            }
+                        }
+                    };
+
+                    let get_cast_op = |s: &Statement<'tcx>| match &s.kind {
+                        StatementKind::Assign(x) => match &x.1 {
+                            Rvalue::Cast(_, op, _) => Some((
+                                x.0,
+                                op.place().expect(
+                                    "Rvalue::Cast is not Operant::Constant so it has a Place",
+                                ),
+                            )),
+                            _ => None,
+                        },
+                        _ => None,
+                    };
+
+                    let find_last_cast_curr_block = || {
+                        bb_data
+                            .statements
+                            .iter()
+                            .rev()
+                            .find_map(|s| get_cast_op(s))
+                            .map(|(casted_to, casted_from)| (casted_from, casted_to))
```

Comment:
This seems like a really confusing way to implement this.  For one thing, it means that the `casted` and `castee` variables above could either be the source and destination of the cast, or they could be the destination and the source.

Reply:
I had a follow-up PR to this, #826, which does I think clarify this part of the code with `struct CVoidCast`, if you think that's a clearer way to organize this logic.

Reply:
I merged #826